source/robot_model/src/Model.cpp

```diff
@@ -29,17 +39,175 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+
+    this->init_geom_model();
+  }
+  else {   // build the model without the package resolver
+    pinocchio::urdf::buildModel(this->get_urdf_path(), this->robot_model_);
+  }
+  
+  // robot data
   this->robot_data_ = pinocchio::Data(this->robot_model_);
+
   // get the frames
   std::vector<std::string> frames;
   for (auto& f : this->robot_model_.frames) {
     frames.push_back(f.name);
   }
   // remove universe and root_joint frame added by Pinocchio
   this->frames_ = std::vector<std::string>(frames.begin() + 2, frames.end());
-  this->init_qp_solver();
+  this->init_qp_solver();  
+}
+
+// Method to initialize collision geometries
+void Model::init_geom_model() {
+    //print building geom model
+
+    pinocchio::urdf::buildGeom(this->robot_model_, 
+                               this->urdf_, 
+                               pinocchio::COLLISION, 
+                               this->geom_model_, 
+                               this->package_paths_);
+    this->geom_model_.addAllCollisionPairs();
+  
+    std::vector<pinocchio::CollisionPair> excluded_pairs = this->generate_joint_exclusion_list();
+
+    // remove collision pairs for linked joints (i.e. parent-child joints)
+    for (const auto& pair : excluded_pairs) {
+        this->geom_model_.removeCollisionPair(pair);
+    }
+
+    this->geom_data_ = pinocchio::GeometryData(this->geom_model_);
+}
+
+std::vector<pinocchio::CollisionPair> Model::generate_joint_exclusion_list() {
+    std::vector<pinocchio::CollisionPair> excluded_pairs;
+    // Iterate through all joints, except the universe joint (0), which has no parent
+    for (pinocchio::JointIndex joint_id = 1u; joint_id < static_cast<pinocchio::JointIndex>(this->robot_model_.njoints); ++joint_id){
+        // Find the parent joint of the current joint
+        pinocchio::JointIndex parent_id = this->robot_model_.parents[joint_id];
+        
+        // TODO: Replace this logic with actual geometry index lookup
+        auto getGeometryIndexForJoint = [](pinocchio::JointIndex joint_id) -> int {
+            return static_cast<int>(joint_id);
+        };
+
+        int geometryIndex1 = getGeometryIndexForJoint(joint_id);
+        int geometryIndex2 = getGeometryIndexForJoint(parent_id);
+
+        // Check if the geometry indices are not equal
+        if (geometryIndex1 != geometryIndex2) {
+            excluded_pairs.push_back(pinocchio::CollisionPair(geometryIndex2, geometryIndex1));
+        }
+    }
+    return excluded_pairs;
+}
+
+unsigned int Model::get_number_of_collision_pairs() {
+  if (this->load_collision_geometries_){
+    return this->geom_model_.collisionPairs.size();
+  }
+  return 0;
+}
+
+
+bool Model::is_geometry_model_initialized() {
+  if (this->load_collision_geometries_){
+    return !this->geom_model_.collisionPairs.empty();
+  }
+  return false;
+}
+
+// Collision detection method
+bool Model::check_collision(const state_representation::JointPositions& joint_positions) {
```

Comment:
it seems to me that you should check `is_geometry_model_initialized()` here before proceeding. I would even suggest that you raise an exception if there is no geometry model (rather than implicitly return false or something like that).

For that you can define a new exception class under `robot_model::exceptions::CollisionGeometryException` or similar.